sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala

```diff
@@ -40,6 +41,104 @@ import org.apache.spark.sql.sources.{DataSourceStrategy, BaseRelation, DDLParser
 
 /**
  * :: AlphaComponent ::
+ * The entry point for parsing the SQL text to logical plan, we call this SQL Dialect.
+ * Currently we support the default dialect named "sql", associated with the class
+ * [[DefaultSQLDialect]]
+ *
+ * And we can also provide custom SQL Dialect, register it before using.
+ * For example, assume we need to provide a Hive Dialect for HiveContext, we can
+ * register the Dialect during the HiveContext creation
+ * {{{
+ *   class HiveContext extends SQLContext {
+ *   ...
+ *   SQLDialectFactory.registerSQLDialect("hiveql", classOf[HiveQLDialect])
+ *   ...
+ *   }
+ * }}}
+ *
+ * And in Spark SQL CLI:
+ * {{{
+ *   -- switch to "hiveql" dialect
+ *   spark-sql>SET spark.sql.dialect=hiveql;
+ *   spark-sql>SELECT * FROM src LIMIT 1;
+ *   -- switch to "sql" dialect
+ *   spark-sql>SET spark.sql.dialect=sql;
+ *   spark-sql>SELECT * FROM src LIMIT 1;
+ * }}}
+ */
+@AlphaComponent
+abstract class SQLDialect {
+  /**
+   * We assume the DDLParser has higher priority than any of the other SQL Parsers,
```

Comment:
I agree with @scwf 
Since our goal is to support variety sql dialects, we can not expect them all have the same behaviours so that the priority of parser is a problem.
What about leave each dialect's own implementation and abstract a method in `SQLDialect `  to let each dialect implement their own order of parsing ?